platform/mbed_application.c

```diff
@@ -0,0 +1,141 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <stdarg.h>
+#include "device.h"
+#include "platform/mbed_application.h"
+#include "platform/mbed_assert.h"
+#include "platform/toolchain.h"
+#include "platform/critical.h"
+
+#if MBED_APPLICATION_SUPPORT
+
+static void powerdown_nvic(void);
+static void powerdown_scb(uint32_t vtor);
+static void start_new_application(void *sp, void *pc);
+
+void mbed_start_application(uintptr_t address)
+{
+
+    void *sp;
+    void *pc;
+
+    // Interrupts are re-enabled in start_new_application
+    core_util_critical_section_enter();
```

Comment:
is there anywhere exit called? I can see interrupts are enabled in the asm functions below. I am asking as critical sections can be nested (contain counters), thus this would lead that there's currently still one active critical section?